src/audio/copier.c

```diff
@@ -605,18 +626,109 @@ static int copier_params(struct comp_dev *dev, struct sof_ipc_stream_params *par
 	return ret;
 }
 
+static int copier_set_sink_fmt(struct comp_dev *dev, int cmd, void *data,
+			       int max_data_size)
+{
+	struct ipc4_copier_config_set_sink_format *sink_fmt;
+	struct copier_data *cd = comp_get_drvdata(dev);
+	struct ipc4_audio_format *fmt;
+	int buf_size;
+
+	sink_fmt = (struct ipc4_copier_config_set_sink_format *)data;
+	dcache_invalidate_region(sink_fmt, sizeof(*sink_fmt));
+
+	if (max_data_size < sizeof(*sink_fmt)) {
+		comp_err(dev, "error: max_data_size %d should be bigger than %d", max_data_size,
+			 sizeof(*sink_fmt));
+		return -EINVAL;
+	}
+
+	if (sink_fmt->sink_id >= IPC4_COPIER_MODULE_OUTPUT_PINS_COUNT) {
+		comp_err(dev, "error: sink id %d is out of range", sink_fmt->sink_id);
+		return -EINVAL;
+	}
+
+	fmt = &sink_fmt->sink_fmt;
+	buf_size = ipc4_buffer_size(fmt->sampling_frequency, fmt->channels_count, fmt->depth,
+				    IPC4_COPIER_MODULE_SCH_BUF_COUNT);
+
+	if (!cd->out_buf[sink_fmt->sink_id]) {
+		cd->out_buf[sink_fmt->sink_id] = create_buffer(dev, &sink_fmt->sink_fmt,
+							       dev->direction, sink_fmt->sink_id,
+							       buf_size);
+	} else if (cd->out_buf[sink_fmt->sink_id]->stream.size < buf_size) {
+		buffer_free(cd->out_buf[sink_fmt->sink_id]);
+		cd->out_buf[sink_fmt->sink_id] = create_buffer(dev, &sink_fmt->sink_fmt,
+							       dev->direction, sink_fmt->sink_id,
+							       buf_size);
+	}
+
+	cd->converter[sink_fmt->sink_id] = get_converter_func(&sink_fmt->source_fmt,
+							      &sink_fmt->sink_fmt);
+
+	return 0;
+}
+
+static int copier_bind_unbind(struct comp_dev *dev, bool bind, void *data, int max_data_size)
+{
+	struct copier_data *cd = comp_get_drvdata(dev);
+	struct ipc4_module_bind_unbind *bu;
+
+	bu = (struct ipc4_module_bind_unbind *)data;
+	if (max_data_size < sizeof(*bu)) {
+		comp_err(dev, "error: max_data_size %d should be bigger than %d", max_data_size,
+			 sizeof(*bu));
+		return -EINVAL;
+	}
+
+	/* copier has only one input and 4 output buffers, only
+	 * processes output buffers when it is a source module
+	 */
+	if (dev->ipc_config.id == IPC4_COMP_ID(bu->header.r.module_id, bu->header.r.instance_id)) {
+		if (!bind) {
+			buffer_free(cd->out_buf[bu->data.r.src_queue]);
+			cd->out_buf[bu->data.r.src_queue] = NULL;
+		} else if (!cd->out_buf[bu->data.r.src_queue]) {
+			int queue = bu->data.r.src_queue;
+
+			comp_err(dev, "error: src queue %d doesn't exist", queue);
+			return -EINVAL;
+		}
+	}
+
+	return 0;
+}
+
+static int copier_cmd(struct comp_dev *dev, int cmd, void *data,
```

Comment:
Can we bind with the source/sink lists today.

Reply:
we create buffer in each modules based on module config, so I want to maintain buffer in each module, not in ipc message processing function